### ISSUE TITLE
server/headless: pure virtual method called with --gdnative-generate-json-api

### ISSUE THREAD

**Issue by lyuma** [godotengine/godot#36582]
**Godot version:**
3.2-stable

**OS/device including version:**
Linux
$ lsb_release -r
Release:	18.04
$ uname -a
Linux server 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Issue description:**
Running a platform=server build with --gdnative-generate-json-api results in a "pure virtual method called" segfault in AudioDriverDummy::thread_func()

```
$ bin/godot_server.x11.opt.tools.64 --gdnative-generate-json-api api.json
Godot Engine v3.2.stable.custom_build.4e7d75ccd - https://godotengine.org
 
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
```

GDB stacktrace:
```
pure virtual method called
terminate called without an active exception

Thread 4 "godot_server.x1" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff7e64700 (LWP 3900)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff6cc2801 in __GI_abort () at abort.c:79
#2  0x00007ffff76b5957 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff76bbab6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff76bbaf1 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff76bc8bf in __cxa_pure_virtual () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00005555577dbabf in AudioDriverDummy::thread_func () at servers/audio/audio_driver_dummy.cpp:68
#7  0x000055555614b3ec in ThreadPosix::thread_callback () at drivers/unix/thread_posix.cpp:74
#8  0x00007ffff7bbd6db in start_thread (arg=0x7ffff7e64700) at pthread_create.c:463
#9  0x00007ffff6da388f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

**Steps to reproduce:**
1. git checkout 3.2-stable
2. /usr/bin/scons platform=server target=release_debug -j4
3. bin/godot_server.x11.opt.tools.64 --gdnative-generate-json-api api.json

**Minimal reproduction project:**
N/A

**Comment by IronicallySerious** [comment]
Could be caused by #36367 but it does really look like that from the stack trace. The PR removes driver/dummy/audio/driver_dummy.h which isn't being used anywhere. The stack trace points to servers/audio/audio_driver_dummy.h instead

**Comment by akien-mga** [comment]
@IronicallySerious that PR is not in 3.2-stable, so it's something else.

**Comment by lyuma** [comment]
There is a AudioDriverDummy declared in static initialization as part of AudioDriverManager in servers/audio_server.h:
`static AudioDriverDummy AudioDriverManager::dummy_driver;`

The init function is called from:
```
(gdb) bt
#0  AudioDriverDummy::init () at servers/audio/audio_driver_dummy.cpp:36
#1  0x00005555575a43fc in AudioDriverManager::initialize () at servers/audio_server.cpp:191
#2  0x00005555557aa5f0 in OS_Server::initialize () at platform/server/os_server.cpp:91
#3  0x00005555557c59bb in Main::setup2 () at main/main.cpp:1201
#4  0x00005555557cb2ff in Main::setup () at main/main.cpp:1146
#5  0x00005555557a66b4 in main () at platform/server/godot_server.cpp:38
```

The destructor is called from:
```
(gdb) bt
#0  AudioDriverDummy::~AudioDriverDummy () at servers/audio/audio_driver_dummy.cpp:134
#1  0x00007ffff6cc5041 in __run_exit_handlers (status=0, listp=0x7ffff706d718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, 
    run_dtors=run_dtors@entry=true) at exit.c:108
#2  0x00007ffff6cc513a in __GI_exit (status=<optimized out>) at exit.c:139
#3  0x0000555555caf322 in NativeScriptLanguage::init () at modules/gdnative/nativescript/nativescript.cpp:1085
#4  0x0000555557ba0b03 in ScriptServer::init_languages () at core/script_language.cpp:177
#5  0x00005555557c6af0 in Main::setup2 () at main/main.cpp:1364
#6  0x00005555557cb2ff in Main::setup () at main/main.cpp:1146
#7  0x00005555557a66b4 in main () at platform/server/godot_server.cpp:38
```

Hence, this pure virtual function call is a use-after-free caused by the thread continuing to run during the C++ static destruction phase.

Because `exit()` is not being called from a conventional place, it means that ::finish() is never called on any components that were previously ::init()ed. The destructor does not clean up the created threads, so this could be a plain old race condition.

I am able to avert this crash with the following workaround:
```
--- servers/audio/audio_driver_dummy.cpp	2020-02-26 22:21:54.194735819 +0000
+++ servers/audio/audio_driver_dummy.cpp	2020-02-27 13:00:48.436284974 +0000
@@ -132,5 +132,5 @@
 };
 
 AudioDriverDummy::~AudioDriverDummy(){
-
+	finish();
 };
```

However, I do not wish to submit this as a PR, as this goes against the existing style for servers, where deallocation of resources is done inside ::finish() rather than the destructor. It is also generally frowned upon to do cleanup of static objects in the destructor because the order is unspecified and may differ across machines or platforms.

**Comment by akien-mga** [comment]
There's already code in `AudioDriver::finish()` that takes care of calling `finish()` on each driver:
https://github.com/godotengine/godot/blob/ac446570d6cb24c4fdb99f53396902c817b46822/servers/audio_server.cpp#L1072-L1076

So not sure while it's not called/not sufficient here.

**Comment by kuruk-mm** [comment]
> There's already code in `AudioDriver::finish()` that takes care of calling `finish()` on each driver:
> https://github.com/godotengine/godot/blob/ac446570d6cb24c4fdb99f53396902c817b46822/servers/audio_server.cpp#L1072-L1076
> 
> So not sure while it's not called/not sufficient here.

Seems the destructor of AudioDriverDummy is beeing called before this.

As @lyuma it's being destroyed when static objects are deleted

Maybe dummy_driver is should be an singleton and non static?

**Comment by kuruk-mm** [comment]
It's working, what do you think? This is the right solution?

https://github.com/kuruk-mm/godot/commit/77336ff4f138ebaccc6d3d70b098a4481b3fb7cc

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

**Comment by luispadron** [comment]
@KoBeWi Yes, I'm on a custom build `v3.2.4.beta.custom_build.7d45e6764` and when using the server build and running the `--gdnative-generate-json-api` i get the same error as OP

**Comment by luispadron** [comment]
Looks like also reported more recently in #43953

**Comment by luispadron** [comment]
Just tested @kuruk-mm's fix and while it does fix the virtual function crash the command does not seem to generate anything. Instead it simply runs godot.

Output:

```sh
./bin/godot_server.osx.opt.debug.64 --gdnative-generate-json-api api.json
Godot Engine v3.2.4.beta.custom_build.1f5bf8551 - https://godotengine.org

```

Running outside of a directory with a project file produces this:

```
./bin/godot_server.osx.opt.debug.64 --gdnative-generate-json-api api.json
Error: Couldn't load project data at path ".". Is the .pck file missing?
If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).
ALERT: ALERT!: Error: Couldn't load project data at path ".". Is the .pck file missing?
If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).
```

Is `--gdnative-generate-json-api` not intended for headless/server builds? I ask because I think it would make dockerizing a server/headless build with the Godot C++ module a lot simpler since that module requires generating the latest `api.json`. Which only seems possible with a normal build of godot.

**Comment by Bromeon** [comment]
This still seems to be a problem for Godot versions `<= 3.3`.
I haven't experienced the issue for `>= 3.3.1`, but I haven't tested very extensively, either.
Could it be that this has been fixed in the meantime (maybe unknowingly)?

We discovered the issue during CI pipelines based on GitHub Actions, which use the headless version of Godot. The pure virtual method call occurs only sporadically, but this makes each CI run non-deterministic.
